My Site Dashboard: Page Behaviours - Loading States - Part 1

ashiagr opened the pull request:
Parent #15625

Page Behavior (pbArwn-3ny-p2)

This PR displays
- loading indicator on initial site load, 
- no loading indicator when there's no site.

Loading on pull-to-refresh was already functional, so, below `loading states` can be tested with this PR:

State (1) - No data, loading (without shimmer - see p1639483178142800-slack-C0290FLA0RM)
State (2) - Finished loading
State (6) - Refresh or loading with content

Yet to handle:
State (7 + 8) - Refresh fails

## To test

Test loading indicator visibility on initial app load, pull-to-refresh, and site switch for an account with and without sites as follows:

MySiteDashboardPhase2FeatureConfig | With Site | Without Sites 
-- | -- | --
OFF | No | No
ON | Yes | No

## Regression Notes
1. Potential unintended areas of impact
Refresh does not finish and the loading indicator keeps appearing.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested for site + no site scenarios. 

3. What automated tests I added (or what prevented me from doing so)
Fixed existing unit tests in different `my site sources` for the initial loading indicator visibility changes.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Set initial refresh value to true

This is done to show a refresh indicator on initial load.

ashiagr committed:
Initiate refresh on my site fragment create

ashiagr committed:
Send postState(CardsUpdate(result)) from CardsSource after cards are fetched

This is done so that the initial loading indicator is not dismissed too soon while the cards are updated from the local cache. Once the fetch is done, refresh is set to false to dismiss the loading indicator.

ashiagr committed:
Handle no site scenario

- For site dependent sources, initial refresh value now depends on whether a site is present.
- For site independent source (currently only CurrentAvatarSource), onRefreshed is explicitly called to immediatley refresh value to false once refresh is complete.
- From MySiteSourceManager, source is refreshed only when it is site independent or when a site is present.

ashiagr committed:
Tests: Fix DomainRegistrationSource existing tests

Set site before DomainRegistrationSource init

ashiagr committed:
Tests: Fix MySiteSourceManager existing tests

ashiagr committed:
Tests: Fix CardsSource existing tests

Set site before CardsSource init

ashiagr committed:
Tests: Fix SelectedSiteSource existing tests

Also extracted out SelectedSiteSource and created from tests with/ without site

ashiagr committed:
Tests: Fix DynamicCardsSource existing tests

Also extracted out DynamicCardsSource and created from tests with/ without site

ashiagr committed:
Tests: Add MySiteSourceManager tests for "no site" scenario

- Site independent sources should be refreshed
- Site dependent sources should not be refreshed

ashiagr committed:
Tests: (Refactor) Extract out ScanAndBackupSource and create from tests with/ without site

ashiagr committed:
Tests: (Refactor) Set site before QuickStartCardSource source init

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsSource.kt:
```diff
@@ -78,9 +81,14 @@ class CardsSource @Inject constructor(
+            val error = result.error
+            when {
+                error != null -> postState(CardsUpdate(result))
+                model != null -> refresh.postValue(false)
```
Suggestion:How about adding a new method (or adjusting) the onRefreshed() method in the interface, so that refresh is never directly updated? This was the reason that onRefreshed was added originally. Wdyt?

ashiagr replied:
Nice idea! `onRefreshed()` adjusted in c1589e8 and used only interface methods to update `refresh` value.

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteFragment.kt:
```diff
@@ -559,9 +560,7 @@ class MySiteFragment : Fragment(R.layout.my_site_fragment),
     }
 
     private fun hideRefreshIndicatorIfNeeded() {
-        if (swipeToRefreshHelper.isRefreshing) {
```
❓ ... so I understand, the guard was eliminated because we start with true (or siteSelected.hasSite) now? We don't need to protect showing the indicator when it shouldn't be showing, correct?

ashiagr replied:
That's what I inferred too. I also noticed that `swipeToRefreshHelper` is enabled only `on resume` and if it is disabled, `swipeToRefreshHelper.isRefreshing` check does not allow the indicator to display, the value being false.  This was the main reason I removed the condition,  let me know if you foresee any issues. 😄

zwarm replied:
👍  Yes, we are thinking the same thing. Thanks!!

ashiagr committed:
Adjust onRefreshed and modify refresh using only the interface methods

ashiagr committed:
Merge branch 'develop' into issue/15625-page-behaviours-loading-part-i

ashiagr committed:
Split onRefreshed to update refresh state based on thread it is accessed

ashiagr committed:
Remove unused import

zwarm reviewed (approved):
Thanks @ashiagr for the changes to onRefreshed() 🙇  Nice work on all the test refactoring - you had to touch every source test to set the site before any source kickoff. 🚀 

I have one ❓  about showing the indicator on initial load. I am not seeing it. I don't know if I should be or if my data is loading too quickly for it to be seen. 

**Steps**
- Fresh install (or clear app data)
- Launch App & Login
- Flip feature flag on
- Close App (no log out)
- Relaunch App 
- Expected to see refreshing indicator, but I did not.

The other tests I ran returned the expected results:  the refreshing indicator is visible when switching sites, when returning to My Site after other navigation, not shown during no site logic and shows correctly when requesting a PTR.

I'm going to approve the PR, but will **not** merge until you have time to answer my question. Thanks.

ashiagr committed:
Set refresh initial value to true

This is done because a site is selected the first time on WPMAinActivity's resume whereas my site sources are created before that. Based on the context we have, rather than selecting the site in the WPMainActivity's onCreate (which is done in a few scenarios like opening from shortcut), it feels more safe to update my site sources' refresh initial value to true.

This means the initial loading indicator will get visible again for no site scenario. This is fixed as part of the next commit.

ashiagr committed:
Check refresh status selectively based on whether a site is selected

When there's no site, it is sufficient to check refreshing status for site independent sources. Other (site dependent) sources have initial refresh value true, checking their refreshing status for no site scenario will provide a wrong status and wrongly display an initial loading indicator.

ashiagr commented:
Thanks for the review @zwarm!

It seems like a site is set in the site repository on [`WPMainActivity.onResume`](https://github.com/wordpress-mobile/WordPress-Android/blob/2eb2a2d6d9696a58d6e0bcced55faad3ab3ab2f1/WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java#L818) whereas my site sources are built before that which might be causing the issue.

I've tried to fix it in 3b63c3d and cc09290, can you give it another try? 🙇‍♀️

zwarm commented:
@ashiagr - Thank you for making the changes. ❤️  I see the refreshing indicator now. I will go ahead and merge. :shipit:
